buttplug/src/server/device/protocol/mizzzee_v2_1.rs

```diff
@@ -0,0 +1,156 @@
+};
+use async_trait::async_trait;
+use std::sync::atomic::{AtomicU32, AtomicU8, Ordering};
+use std::{sync::Arc, time::Duration};
+
+generic_protocol_initializer_setup!(MizzZeeV2_1, "mizzzee-v2.1");
+
+#[derive(Default)]
+pub struct MizzZeeV2_1Initializer {}
+
+#[async_trait]
+impl ProtocolInitializer for MizzZeeV2_1Initializer {
+  async fn initialize(
+    &mut self,
+    hardware: Arc<Hardware>,
+    _: &ProtocolDeviceAttributes,
+  ) -> Result<Arc<dyn ProtocolHandler>, ButtplugDeviceError> {
+    Ok(Arc::new(MizzZeeV2_1::new(hardware)))
+  }
+}
+
+// Time between MizzZee2 update commands, in milliseconds.
+const MIZZZEE2_COMMAND_DELAY_MS: u64 = 20;
+
+// Time between MizzZee2 keep vibrating commands, in milliseconds.
+const MIZZZEE2_COMMANDS_KEEP_VIBRATING: u64 = 200;
+
+// Amount of commands that can be skipped without stopping the device.
+const MIZZZEE2_COMMANDS_TO_SKIP: u8 = (MIZZZEE2_COMMANDS_KEEP_VIBRATING / MIZZZEE2_COMMAND_DELAY_MS - 1) as u8;
+
+fn handle_scale(scale: f32) -> f32 {
+  if scale == 0.0 { return 0.0; }
+  scale * 0.7 + 0.3
+}
+
+fn scalar_to_vector(scalar: u32) -> Vec<u8> {
+  const HEADER: [u8; 3] = [0x03, 0x12, 0xf3];
+  const FILL_VEC: [u8; 6] = [0x00, 0xfc, 0x00, 0xfe, 0x40, 0x01];
+
+  let scale: f32 = handle_scale(scalar as f32 / 1000.0) * 1023.0;
+  let modded_scale: u16 = ((scale as u16) << 6) | 60;
+  
+  let bytes = modded_scale.swap_bytes().to_be_bytes();
+
+  let mut data: Vec<u8> = Vec::new();
+  data.extend_from_slice(&HEADER);
+  data.extend_from_slice(&FILL_VEC);
+  data.extend_from_slice(&bytes);
+  data.extend_from_slice(&FILL_VEC);
+  data.extend_from_slice(&bytes);
+  data.push(0x00);
+
+  data
+}
+
+async fn vibration_update_handler(
+  device: Arc<Hardware>,
+  loops_skipped_holder: Arc<AtomicU8>,
+  last_scalar_holder: Arc<AtomicU32>,
+  current_scalar_holder: Arc<AtomicU32>,
+) {
+  info!("Entering MizzZee2 Control Loop");
+  loop {
+    sleep(Duration::from_millis(MIZZZEE2_COMMAND_DELAY_MS)).await;
+
+    let loops_skipped = loops_skipped_holder.load(Ordering::Relaxed);
+    let last_scalar = last_scalar_holder.load(Ordering::Relaxed);
+    let current_scalar = current_scalar_holder.load(Ordering::Relaxed);
+
+    if last_scalar == current_scalar && loops_skipped < MIZZZEE2_COMMANDS_TO_SKIP {
+      loops_skipped_holder.store(loops_skipped + 1, Ordering::Relaxed);
+      continue;
+    }
+
+    loops_skipped_holder.store(0, Ordering::Relaxed);
+    last_scalar_holder.store(current_scalar, Ordering::Relaxed);
+
+    if device
+      .write_value(&HardwareWriteCmd::new(
+        Endpoint::Tx,
+        scalar_to_vector(current_scalar),
+        true
+      ))
+      .await
+      .is_err() { break; }
+
+    info!("MZ2 scalar: {}", current_scalar);
+  }
+  info!("MizzZee2 control loop exiting, most likely due to device disconnection.");
+}
+
+#[derive(Default)]
+pub struct MizzZeeV2_1 {
+  current_scalar: Arc<AtomicU32>,
+}
+
+impl MizzZeeV2_1 {
+  fn new(device: Arc<Hardware>) -> Self {
```

Comment:
In here would be a great place to query the characteristic to get the multiplier mode value.

Reply:
I just... Don't know how to do that...

Reply:
So...
WeChat applet has something like this:
```js
wx.openBluetoothAdapter({
	mode: "central",
	success: function(t) {
		wx.onBluetoothDeviceFound(onBLEDeviceFound);
		wx.onBLEConnectionStateChange(onBLEConnChange);
		wx.onBLECharacteristicValueChange(onBLECharValueChange)
		e & e(true)
	},
	fail: function(t) {
		e && e(false)
	}
});
```

and

```js
charValueChange = function(t) {
	e.emit("characteristicValueChange", t)
	if ("00002A50-0000-1000-8000-00805F9B34FB" === t.characteristicId.toUpperCase()) {
		var n = new DataView(t.value);
		deviceType = n.getUint16(3, true)
	}
}
```

So it means that it could be characteristic of any service.
`0xFF10` `Unknown Service` has characteristics:
- `0xFF12` (Read, Write)  
  On read returns `0x15`  
  Has descriptor `Client Characteristic Configuration` `0x2902` which returns `Notifications enabled` on read
- `0xF11` (Notify, Read)  
  On read returns `0x15`

<details><summary>Other services</summary>
<p>

`0x1800` `Generic Access`:
- `0x2A00` (Read) - Device Name: `XHTKJ`
- `0x2A01` (Read) - Appearance: `[961] Keyboard (HID subtype)`
- `0x2A04` (Read) - Peripheral Preferred Connection Parameters: `Connection Interval: 7.50ms - 7.50ms,\nSlave Latency: 100,\nSupervision Timeout Multiplier: 300`


`0x1801` `Generic Attribute`:
- `0x2A05` (Indicate) - Service Changed  
  Descriptors:
  - `0x2902` - Client Characteristic Configuration  
    Value: `Indications enabled`

`0x180A` `Device Information`:
- `0x2A23` (Read) - System ID: `0x519daa000019fb50`
- `0x2A24` (Read) - Model Number String: `Model Number`
- `0x2A25` (Read) - Serial Number String: `Serial Number`
- `0x2A26` (Read) - Firmware Revision String: `Firmware Revision`
- `0x2A27` (Read) - Hardware Revision String: `Hardware Revision`
- `0x2A28` (Read) - Software Revision String: `Software Revision`
- `0x2A29` (Read) - Manufacturer Name String: `Manufacturer Name`
- `0x2A2A` (Read) - IEEE 11073-20601 Regulatory Certification: `0xfe006578706572696d656e74619c`
- `0x2A50` (Read) - PnP ID: `Bluetooth SIG Company: Chipsea Technologies (ShenZhen) Corp. <0x06A7>\nProduct Id: 0\nProduct Version: 1`


</p>
</details>

Reply:
So i can't just add `rx: 00002a50-00...` to `FF10` service and read value - this service doesn't have `0x2A50` characteristic

Reply:
So... Now I just wait, I guess...